File: backend/src/inventario/permissions.py
from .models import DEFAULT_SALE_MARGIN_PERCENT


SALE_MARGIN_MODULE = "inventory_sale_margin"
CREATE_ITEMS_MODULE = "inventory_items_create"


def user_has_module(user, module: str) -> bool:
    if not user or not getattr(user, "is_authenticated", False):
        return False
    if getattr(user, "is_superuser", False):
        return True

    profile = getattr(user, "access_profile", None)
    if not profile:
        return False

    return module in profile.normalized_modules()


def can_edit_sale_margin(user) -> bool:
    return user_has_module(user, SALE_MARGIN_MODULE)


def can_create_inventory_items(user) -> bool:
    return user_has_module(user, CREATE_ITEMS_MODULE)


def default_sale_margin_if_unauthorized(value, user):
    if can_edit_sale_margin(user):
        return value or DEFAULT_SALE_MARGIN_PERCENT
    return DEFAULT_SALE_MARGIN_PERCENT

File: frontend/src/pages/Inventory/permissions.ts
import { hasModule } from "../../models/modules";

export const DEFAULT_SALE_MARGIN_PERCENT = "50";
export const CREATE_ITEMS_MODULE = "inventory_items_create";
export const BATCHES_MODULE = "inventory_batches";
export const SALE_MARGIN_MODULE = "inventory_sale_margin";

export function canCreateInventoryItems(user: { is_superuser?: boolean; modules?: string[] } | null | undefined): boolean {
  return hasModule(user, CREATE_ITEMS_MODULE);
}

export function canRegisterBatches(user: { is_superuser?: boolean; modules?: string[] } | null | undefined): boolean {
  return hasModule(user, BATCHES_MODULE);
}

export function canEditSaleMargin(user: { is_superuser?: boolean; modules?: string[] } | null | undefined): boolean {
  return hasModule(user, SALE_MARGIN_MODULE);
}

export function saleMarginOrDefault(value: string | null | undefined, canEdit: boolean): string {
  if (!canEdit) return DEFAULT_SALE_MARGIN_PERCENT;
  return value || DEFAULT_SALE_MARGIN_PERCENT;
}
